src/utils/StorageHandler.ts

```diff
@@ -548,11 +548,19 @@ export default class StorageHandler {
       }
     } else {
       const allTargetApps = await this.getTargetAppNames();
+      const impactedTargetApps = new Set(entity.targetApps);
+      targetApps.forEach(value => allTargetApps.has(value) && impactedTargetApps.add(value));
       await Promise.all(
-        Array.from(allTargetApps).map((targetApp) =>
-          Array.from(targetApps).includes(targetApp)
-            ? this.addEntityAssocs(entityNames, targetApp)
-            : this.removeEntityAssocs(entityNames, targetApp)
+        Array.from(impactedTargetApps).reduce<Promise<void>[]>(
+          (targetAppPromises, targetApp) => {
+            if (targetApps.has(targetApp) && !entity.targetApps.has(targetApp)) {
+              targetAppPromises.push(this.addEntityAssocs(entityNames, targetApp));
+            } else if (!targetApps.has(targetApp) && entity.targetApps.has(targetApp)) {
+              targetAppPromises.push(this.removeEntityAssocs(entityNames, targetApp));
+            }
+            return targetAppPromises;
+          },
+          []
```

Comment:
This is what I had in mind. I think it'd be a lot simpler
```suggestion
        filterNulls(
          Array.from(allTargetApps).map((targetApp) =>
            targetApps.has(targetApp)
              ? this.addEntityAssocs(entityNames, targetApp)
              : entity.targetApps.has(targetApp)
              ? this.removeEntityAssocs(entityNames, targetApp)
              : null
          )
```